### PR TITLE
Spec compliancy of check-es2015-constants plugin

### LINKED ISSUE #5728: Replace "check-es2015-constants" plugin with injected `throw` statements
The "check-es2015-constants" plugin is technically not spec-compliant. Nothing about the spec says that Babel should throw an error at compile-time about assigning to a constant binding.

```
(function(){
  const a = "foo";

  if (false) a = "false";

  return a;
})();
```
for instance is perfectly valid code and will run to completion. It's weird and probably won't come up in real code, but I don't see a reason that Babel should be diverging in behavior here.

There are two cases where this plugin currently applies, to my knowledge

1. Assigning to `const` like the example above
2. Assigning to module bindings like `import foo from "foo"; foo = 4;`

In the first case, I'd vote for the block-scoping transform to automatically rewrite every Assignment/UpdateExpression to insert a `throw` into the output code.

In the `import` case, I'd vote for the module transform to be responsible for injecting `throw` statements into the code.

This doesn't strictly need to happen for Babel 7.x, I don't think, since if anything it is relaxing restrictions that are currently _too_ strict, but it would certainly be a nice-to-have.

This is also a reasonably self-contained task, so it would be a good first task for someone to tackle.

### PR DESCRIPTION
**Pull request by maurobringolf**
| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Deprecations?            | No
| Spec Compliancy?         | Yes
| Tests Added/Pass?        | Yes
| Fixed Tickets            | Fixes #5728 
| License                  | MIT
| Doc PR                   | 
| Dependency Changes       | 

I started working on the proposed change to the check-es2015-constants plugin: Instead of throwing a compile time error when `const` is violated, Babel should insert a throw statement before the violation. This will make it a runtime error and violations in code that is not executed are fine. I added the example from the issue as a test case and changed existing test cases.

I am not sure if this covers all possible violations besides modules. Is there a list of all possible violation types somewhere? Variable declarations via module import statements are not yet covered by this PR. Let me know if this is going into the right direction or if I should take a different route.

### CODE REVIEW

**Comment by maurobringolf** [comment]
Actually I just realized that inserting before the violation will break in many many many cases. I think the throwStatement needs to replace the violation instead.

**Comment by jridgewell** [comment]
Replacing won't work either. You'll need to find the statement parent and insert before that.

**Comment by maurobringolf** [comment]
Agreed, finding a parent statement seems more reasonable. Do you have an example where my current implementation fails? I tried using `path.getStatementParent` and inserting before that. It does not behave correctly in the example from the issue:

**Input**
```javascript
const a = "foo";
if (false) a = "false";
```

**Output**

```javascript
const a = "foo";
throw new Error("\"a\" is read-only");
if (false) a = "false";
```

I checked and indeed, `violation.getStatementParent` returns the `IfStatement` in this case. This is because of the shorthand `IfStatement`. When using curly braces around the violation, it works correctly because then `getStatementParent` will return the actual if clause and not the complete `IfStatement`.

The same happens with all loop constructs that can either have a single expression or a block as body. Do I have to manually work around these or is there a better way than `getStatementParent`?

**Comment by jridgewell** [comment]
```js
const before = path.findParent(p => p.isStatement() || p.parentPath.isArrowExpression());
```

File: packages/babel-plugin-check-es2015-constants/test/fixtures/general/no-for-in/expected.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+var MULTIPLIER = 5;
+throw new Error("\"MULTIPLIER\" is read-only");
+
+for (MULTIPLIER in arr) {}
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-check-es2015-constants/test/fixtures/general/no-for-in/expected.js]: This should be
```js
var MULTIPLIER = 5;

for (MULTIPLIER in arr) {
  throw new Error("\"MULTIPLIER\" is read-only");
}
```

Because if arr is an empty object, `MULTIPLIER` is never reassigned and thus it shouldn't throw.

Comment (maurobringolf) [inline packages/babel-plugin-check-es2015-constants/test/fixtures/general/no-for-in/expected.js]: Ah true, thank you! I will change the `forXStatement` case then.

File: packages/babel-plugin-check-es2015-constants/src/index.js [hunk 2]
```diff
@@ -44,7 +44,15 @@ export default function({ messages, types: t }) {
                 ]),
               );
             } else if (violation.parentPath.isForXStatement()) {
-              violation.parentPath.insertBefore(throwNode);
+              // Transform single statement body into BlockStatement
+              if (!violation.parentPath.get("body").isBlockStatement()) {
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: I think you can use `violation.parentPath.ensureBlock()` (I'm not 100% sure because I've never used it, but I found it while looking at random babel files 😆 )

Comment (maurobringolf) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: Ah cool! As far as I can tell it uses this converter function, so it is probably appropriate:

https://github.com/babel/babel/blob/4e50b2d9d9c376cee7a2cbf56553fe5b982ea53c/packages/babel-types/src/converters.js#L243-L259

I will give it a try, thanks!

File: packages/babel-plugin-check-es2015-constants/src/index.js [hunk 3]
```diff
@@ -7,7 +7,53 @@ export default function({ messages }) {
...
+                t.stringLiteral(messages.get("readOnly", name)),
+              ]),
+            );
+
+            if (violation.isAssignmentExpression()) {
+              violation
+                .get("right")
+                .replaceWith(
+                  t.sequenceExpression([
+                    t.callExpression(
```
Comment (buunguyen) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: Duplicated code with below. Should move to a variable?

File: packages/babel-plugin-check-es2015-constants/src/index.js [hunk 4]
```diff
@@ -7,7 +7,53 @@ export default function({ messages }) {
...
+                        null,
+                        [],
+                        t.blockStatement([throwNode]),
+                      ),
+                      [],
+                    ),
+                    violation.get("right").node,
+                  ]),
+                );
+            } else if (violation.parentPath.isUpdateExpression()) {
```
Comment (buunguyen) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: Is there a test case for this branch?

File: packages/babel-plugin-check-es2015-constants/src/index.js (hunk 3 above)
Comment (maurobringolf) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: Sounds good, I will create one for the comma expression. Should be less lines and easier to follow.

File: packages/babel-plugin-check-es2015-constants/src/index.js (hunk 4 above)
Comment (maurobringolf) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: Currently there is only one in `packages/babel-plugin-check-es2015-constants/test/fixtures/general/update-expression`. Could probably use more since this can show up in a lot of wierd places I think.

Comment (buunguyen) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: Thanks. I'm quite surprised that the check happens at `parentPath` instead of directly on `violation` like the first branch. I guess violations are collected differently for assignment vs update expressions.

Comment (maurobringolf) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: Yes I thought that was confusing as well. Some violations are given as the identifier nodes and some are their parents, but I am not sure why. If these violations are not used anywhere else it might make sense to change it for the sake of consistency.

Comment (buunguyen) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: I think that's a good idea. Maybe it should be addressed in a separate PR and see what other folks feedback.

File: packages/babel-plugin-check-es2015-constants/test/fixtures/general/no-assignment/expected.js [hunk 5]
```diff
@@ -0,0 +1,4 @@
+var MULTIPLIER = 5;
+MULTIPLIER = (function () {
```
Comment (buunguyen) [inline packages/babel-plugin-check-es2015-constants/test/fixtures/general/no-assignment/expected.js]: I'm sure these would pass, but would be great to add test for destructuring assignments, e.g.

```js
const a = 1;
[a] = [2];

const b = 1;
({b} = {b: 2})
```

**Comment by xtuc** [comment]
Is this a breaking change btw? Users won't expect their code to throw at runtime, but atm they wouldn't be able to transpile it anyway.

**Comment by maurobringolf** [comment]
I am not sure if it classifies as breaking change since as you said, runtime errors are only added in cases which did not even get transpiled before. I personally dont see it as a breaking change.

File: packages/babel-plugin-check-es2015-constants/test/fixtures/general/no-assignment/expected.js (hunk 5 above)
Comment (maurobringolf) [inline packages/babel-plugin-check-es2015-constants/test/fixtures/general/no-assignment/expected.js]: Definitely, I want to add executed tests as well to assert the runtime error is thrown. I need to look at how other packages do these tests first though, because everything I have tried so far did not work.

**Comment by xtuc** [comment]
I liked the old behavior, could we add an option `throwAtCompile: Boolean` to restore the error at compile time?

I can make a separate PR.

File: packages/babel-plugin-check-es2015-constants/src/index.js [hunk 6]
```diff
@@ -7,7 +7,32 @@ export default function({ messages }) {
...
           for (const violation of (binding.constantViolations: Array)) {
-            throw violation.buildCodeFrameError(messages.get("readOnly", name));
+            const throwNode = t.throwStatement(
+              t.newExpression(t.identifier("Error"), [
+                t.stringLiteral(messages.get("readOnly", name)),
+              ]),
+            );
+
+            // Returns a comma expression of IIFE wrapped throwNode followed by given expression.
+            const throwBefore = expression =>
```
Comment (jridgewell) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: This can be hoisted outside the visitor by passing in the `throwNode`, too.

Comment (maurobringolf) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: I considered this and decided to keep it inside the visitor for simplicity. But I'm fine with both ways and can change it for sure.

Comment (jridgewell) [inline packages/babel-plugin-check-es2015-constants/src/index.js]: We're redefining a function in a loop, which is super slow.

**Comment by maurobringolf** [comment]
Awesome, thanks I learned a lot!

**Comment by hzoo** [comment]
@xtuc making lots of options kinda sucks so I'd suggest just adding it as "loose" given this is spec

**Comment by xtuc** [comment]
Yeah I agree but isn't it strange to have loose enabled and get an error at compile time?

**Comment by hzoo** [comment]
Yes, it's "weird" but nothing about loose mode says no errors, it literally just means loose about the spec. Otherwise I would just not run the plugin at all?

**Comment by jridgewell** [comment]
Maybe this behavior should be `spec`, and the old throw should be regular?

**Comment by xtuc** [comment]
Yes but spec behavior should be by default. But than what's the difference between `spec: false` and `loose: true` lol

**Comment by loganfsmyth** [comment]
I think http://eslint.org/docs/rules/no-const-assign is a better approach to get a warning right away, since ESLint will even nicely highlight the line for you in your editor.

**Comment by hzoo** [comment]
@xtuc 😄 . I don't think spec: false or loose: false would be a valid option

**Comment by loganfsmyth** [comment]
In the original issue, https://github.com/babel/babel/issues/5728, I tried to say we should split this up into the two plugins that actually need it (block-scoping and modules), but that doesn't appear to have happened in this PR. Would anyone mind if I file a new issue about that?

As it is, if you turn this on for block scoping for instance, you also get it for modules, e.g.

```
import foo from "bar";

foo = "value";
```
with
```
{
  presets: [['es2015', {modules: false}],
}
```

will still inject the logic to throw, even though you turned off processing of modules, and `foo` is not a block-scoped binding. To me it seems like import bindings should be handled by the module transform and `const` bindings should be handled by block-scoping, and we should delete `check-es2015-constants` entirely.

Thoughts?
